flashlib.core.robot/src/main/java/com/flash3388/flashlib/robot/control/PidController.java

```diff
@@ -173,32 +143,34 @@ public void setOutputLimit(double outputLimit){
      * Calculates the output to the system to compensate for the error.
      *
      * @param processVariable the process variable of the system.
-     * @param setPoint the desired set point.
+     * @param setpoint the desired set point.
      *
      * @return the compensation value from the PID loop calculation
      */
-    public double calculate(double processVariable, double setPoint){
+    public double calculate(double processVariable, double setpoint, double currentTimeSeconds) {
         if(mSetPointRange != 0) {
             processVariable = ExtendedMath.constrain(processVariable, processVariable - mSetPointRange, processVariable + mSetPointRange);
         }
 
-        double error = setPoint - processVariable;
+        double error = setpoint - processVariable;
 
         double pOut = mKp.getAsDouble() * error;
-        double fOut = mKf.getAsDouble() * setPoint;
+        double fOut = mKf.getAsDouble() * setpoint;
 
         if(mIsFirstRun){
             mIsFirstRun = false;
             mLastOutput = pOut + fOut;
-            mLastProcessVariable = processVariable;
+            mLastError = error;
+            mLastTimeSeconds = currentTimeSeconds - 1;
         }
 
         double iOut = mKi.getAsDouble() * mTotalError;
-        double dOut = -mKd.getAsDouble() * (processVariable - mLastProcessVariable);
+        double dOut = mKd.getAsDouble() * ((error - mLastError)/(currentTimeSeconds-mLastOutput));
```

Comment:
`currentTimeSeconds-mLastOutput`? Maybe `- mLastTimeSeconds`?

Reply:
yeah, with `Time` that wouldn't have happened